src/compute-client/src/plan/transform/relax_must_consolidate.rs

```diff
@@ -9,3 +9,75 @@
 
 //! Placeholder module for an [crate::plan::transform::Transform] that infers
 //! physical monotonicity.
+
+use std::marker::PhantomData;
+
+use mz_repr::GlobalId;
+
+use crate::plan::interpret::{PhysicallyMonotonic, SingleTimeMonotonic};
+use crate::plan::reduce::{HierarchicalPlan, MonotonicPlan};
+use crate::plan::top_k::{MonotonicTop1Plan, MonotonicTopKPlan, TopKPlan};
+use crate::plan::transform::{BottomUpTransform, TransformConfig};
+use crate::plan::{Plan, ReducePlan};
+
+/// A transformation that takes the result of single-time physical monotonicity
+/// analysis and refines, as appropriate, the setting of the `must_consolidate`
+/// flag in monotonic `Plan` nodes with forced consolidation.
+#[derive(Debug)]
+pub struct RelaxMustConsolidate<T = mz_repr::Timestamp> {
+    _phantom: PhantomData<T>,
+}
+
+impl<T> RelaxMustConsolidate<T> {
+    pub fn new() -> Self {
+        RelaxMustConsolidate {
+            _phantom: Default::default(),
+        }
+    }
+}
+
+impl<'a, T> BottomUpTransform<'a, GlobalId, T> for RelaxMustConsolidate<T> {
+    type Info = PhysicallyMonotonic;
+
+    type Interpreter = SingleTimeMonotonic<'a, T>;
+
+    fn name(&self) -> &'static str {
+        "must_consolidate relaxation"
+    }
+
+    fn interpreter(config: &'a TransformConfig<GlobalId>) -> Self::Interpreter {
+        SingleTimeMonotonic::new(config)
+    }
+
+    fn action(plan: &mut Plan<T>, plan_info: &Self::Info, _input_infos: &[Self::Info]) {
+        // Look at `plan_info` and type of `Plan` node and refine the `must_consolidate` flag.
+        match (plan, plan_info) {
+            (
+                Plan::Reduce {
+                    plan:
+                        ReducePlan::Hierarchical(HierarchicalPlan::Monotonic(MonotonicPlan {
+                            must_consolidate,
+                            ..
+                        })),
+                    ..
+                }
+                | Plan::TopK {
+                    top_k_plan:
+                        TopKPlan::MonotonicTop1(MonotonicTop1Plan {
+                            must_consolidate, ..
+                        }),
+                    ..
+                }
+                | Plan::TopK {
+                    top_k_plan:
+                        TopKPlan::MonotonicTopK(MonotonicTopKPlan {
+                            must_consolidate, ..
+                        }),
+                    ..
+                },
+                PhysicallyMonotonic(true),
```

Comment:
So `must_consolidate` refers to the output and not the input of the node?

Reply:
You are absolutely right -- I should have read my own docs! :-)

Reply:
Maybe we can prevent this confusion by renaming `must_consolidate` to `consolidate_input`, idk?

Reply:
Not sure if this would create some more confusion... the issue is that we have coarse LIR physical operators compared to more fine-grained DD operators. So the consolidation that happens, e.g., in `Plan::Reduce` is not on the input, but rather happening after we apply `KeyValPlan` (in fact, after we do that and try out a "cheap" streaming consolidation). The flag is more or less saying "consolidation must happen inside this black box somewhere" rather than trying to specify whether inputs/outputs to it will actually be consolidated.

Reply:
OK, so probably just copy-pasting the above comment somewhere on top of the flat definition should be sufficient.

Reply:
Cool, I just revised the documentation near the definition of the flags to bring this out.